mixer/adapter/svcctrl/reportbuilder_test.go

```diff
@@ -0,0 +1,228 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"testing"
+	"time"
+
+	sc "google.golang.org/api/servicecontrol/v1"
+
+	"istio.io/istio/mixer/adapter/svcctrl/template/svcctrlreport"
+)
+
+func getTestReportBuilder() *reportBuilder {
+	requestTime, _ := time.Parse(time.RFC3339Nano, "2017-10-21T17:09:05.000Z")
+	responseTime, _ := time.Parse(time.RFC3339Nano, "2017-10-21T17:09:05.100Z")
+
+	return &reportBuilder{
+		supportedMetrics: []metricDef{
+			{
+				name:           "test_metric",
+				valueGenerator: generateRequestCount,
+				labels: []string{
+					"/consumer_id",
+					"/credential_id",
+					"/error_type",
+					"/protocol",
+					"/response_code",
+					"/response_code_class",
+					"/status_code",
+				},
+			},
+		},
+		instance: &svcctrlreport.Instance{
+			ApiVersion:      "v1.0",
+			ApiOperation:    "POST /echo",
+			ApiProtocol:     "REST",
+			ApiService:      "echo.test.com",
+			ApiKey:          "test_key",
+			RequestTime:     requestTime,
+			RequestMethod:   "POST",
+			RequestPath:     "echo.test.com/echo",
+			RequestBytes:    10,
+			ResponseTime:    responseTime,
+			ResponseCode:    200,
+			ResponseBytes:   1024,
+			ResponseLatency: 1 * time.Microsecond,
+		},
+		resolver: &mockConsumerProjectIDResolver{
+			consumerProjectID: "test_consumer_project",
+		},
+		labelMap: make(map[string]string),
+	}
+}
+
+func TestBuildLogEntry(t *testing.T) {
+	rb := getTestReportBuilder()
+	op := &sc.Operation{}
+	rb.addLogEntry(op)
+
+	if len(op.LogEntries) != 1 {
+		t.Errorf(`len(op.LogEntries) != 1`)
+	}
+
+	expected :=
+		`{
+			"api_name":"echo.test.com",
+			"api_method":"POST /echo",
+			"api_key":"test_key",
+			"http_method":"POST",
+			"request_size_in_bytes":10,
+			"http_response_code":200,
+			"timestamp":"2017-10-21T17:09:05Z",
+			"location":"global",
+			"log_message":"Method:POST /echo"
+		}`
+
+	actual := string(op.LogEntries[0].StructPayload)
+	if !compareJSON(expected, actual) {
+		t.Errorf("expect payload %v, but get %v", expected, actual)
+	}
+}
+
+func TestBuildMetricValue(t *testing.T) {
+	rb := getTestReportBuilder()
+	op := &sc.Operation{}
+	rb.addMetricValues(op)
+	expected :=
+		`{
+			"labels":{
+				"cloud.googleapis.com/location":"global",
+				"serviceruntime.googleapis.com/api_method":"POST /echo",
+				"serviceruntime.googleapis.com/api_version":"v1.0",
+				"serviceruntime.googleapis.com/consumer_project":"test_consumer_project"
+			},
+			"metricValueSets":[
+				{
+					"metricName":"test_metric",
+					"metricValues":[
+						{
+							"endTime":"2017-10-21T17:09:05.1Z",
+							"int64Value":"1",
+							"labels":{
+								"/consumer_id":"api_key:test_key",
```

Comment:
Should we keep all labels at the operations level. This way multiple metric values can share the same label. We can do this because metric value's label defaults to the operation label if not specified.